### PR TITLE
Create webpack 4 plugin to emulate fancy webpack builds in wrangler 2

### PR DESCRIPTION
**Pull request by caass**
For users that

- use `type = webpack`
- provide their own webpack config
- do non-trivial modifications to their worker source

this webpack plugin attempts to emulate wrangler 1's behavior

### CODE REVIEW

File: packages/wranglerjs-compat-webpack-plugin/package.json
Comment (petebacondarwin) [inline packages/wranglerjs-compat-webpack-plugin/package.json:17]: I don't think you need this because of the wonders of npm workspaces.

File: packages/wranglerjs-compat-webpack-plugin/src/index.ts
Comment (JacobMGEvans) [inline packages/wranglerjs-compat-webpack-plugin/src/index.ts:198]: why not  `execaSync`?

Comment (caass) [reply packages/wranglerjs-compat-webpack-plugin/src/index.ts:198]: it causes problems with webpack 4, since execa is an ESModule and webpack.config.js needs any plugins to be CommonJS Modules

File: packages/wranglerjs-compat-webpack-plugin/package.json
Comment (caass) [reply packages/wranglerjs-compat-webpack-plugin/package.json]: 👀

**Review by petebacondarwin — APPROVED** [review]
As I am tired, it might be worth getting a second opinion from @threepointone and @JacobMGEvans. But it looks totally awesome!

File: docs/eject-webpack.md
Comment (threepointone) [inline docs/eject-webpack.md:94]:
```suggestion
3. Remove unsupported entries from your `wrangler.toml`
```

File: packages/wranglerjs-compat-webpack-plugin/package.json
Comment (threepointone) [inline packages/wranglerjs-compat-webpack-plugin/package.json:44]:
```suggestion
```
`peerDependencies` behave pretty poorly in newer versions of  npm, and it's not that useful for plugins anyway. let's just remove this.

File: packages/wranglerjs-compat-webpack-plugin/package.json
Comment (threepointone) [inline packages/wranglerjs-compat-webpack-plugin/package.json:55]:
```suggestion
```
I don't think we need this in this package? might want to have a look at trimming `transformIgnorePatterns` as well

File: packages/wranglerjs-compat-webpack-plugin/src/__tests__/helpers/compare-outputs.ts
Comment (threepointone) [inline packages/wranglerjs-compat-webpack-plugin/src/__tests__/helpers/compare-outputs.ts:240]: won't files with same names in different dirs clash here? should this be `dir + entry`? (I'm not actually sure, hence asking)

File: packages/wranglerjs-compat-webpack-plugin/src/__tests__/index.spec.ts
Comment (threepointone) [inline packages/wranglerjs-compat-webpack-plugin/src/__tests__/index.spec.ts:76]: wanna replace these TODO tests with an `it.todo`?

**Review by threepointone — APPROVED** [review]
This PR is pretty fkin incredible. Approving with some questions.

File: packages/wrangler/src/entry.ts
Comment (caass) [reply packages/wrangler/src/entry.ts:93]: Yeah, I'll add some inline snapshots to the test to confirm it

File: packages/wranglerjs-compat-webpack-plugin/package.json
Comment (caass) [reply packages/wranglerjs-compat-webpack-plugin/package.json:55]: Since we depend on `wrangler`, we actually need all of these :(

File: packages/wranglerjs-compat-webpack-plugin/src/__tests__/helpers/compare-outputs.ts
Comment (caass) [reply packages/wranglerjs-compat-webpack-plugin/src/__tests__/helpers/compare-outputs.ts:240]: ummm...i don't think so? Because they'll be like
```
- parent
  | - child
  | - child2
    | - child
```
should become
```
{
  child: <Buffer>
  child2: {
    child: <Buffer>
  }
}
```
right?

File: packages/wrangler/src/entry.ts
Comment (threepointone) [reply packages/wrangler/src/entry.ts:93]: So, this actually hides logs, that's why it needs to be "inherit". Open to alternate solutions, but pending that, a fix in https://github.com/cloudflare/wrangler2/pull/838

**Pull request merged**
